### PR TITLE
Add initialConnections to requestState

### PR DESCRIPTION
**Pull request by david0xd**
This PR adds `initialConnections` specification from manifest to `requestState` object which is used on Snap installation or update process.

Related ticket: https://github.com/MetaMask/MetaMask-planning/issues/1895
Blocking: https://github.com/MetaMask/MetaMask-planning/issues/1895

### CODE REVIEW

File: packages/snaps-controllers/src/snaps/SnapController.ts [hunk 1]
```diff
@@ -2689,6 +2693,9 @@ export class SnapController extends BaseController<
 
       this.#updateApproval(pendingApproval.id, {
         loading: false,
+        ...(initialConnections !== undefined && initialConnections !== null
+          ? { connections: initialConnections }
+          : {}),
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/SnapController.ts:2698]:
```suggestion
        initialConnections: snap.initialConnections ?? {}
```

Wouldn't it be simpler to do something like this?

File: packages/snaps-controllers/src/snaps/SnapController.ts [hunk 2]
```diff
@@ -2231,6 +2231,10 @@ export class SnapController extends BaseController<
         this.#calculatePermissionsChange(snapId, processedPermissions);
 
       this.#updateApproval(pendingApproval.id, {
+        ...(manifest.initialConnections !== undefined &&
+        manifest.initialConnections !== null
+          ? { connections: manifest.initialConnections }
+          : {}),
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/SnapController.ts:2237]:
```suggestion
        initialConnections: snap.manifest?? {}
```
Wouldn't it be simpler to do something like this?

File: packages/snaps-utils/src/snaps.ts [hunk 3]
```diff
@@ -85,6 +89,10 @@ export type PersistedSnap = Snap;
  * A Snap as it exists in {@link SnapController} state.
  */
 export type Snap = TruncatedSnap & {
+  /**
+   * The initial connections of the Snap, optional, requested on installation.
+   */
+  initialConnections?: InitialConnections;
```
Comment (FrederikBolding) [inline packages/snaps-utils/src/snaps.ts:95]: This type is not correct unless you add `initialConnections` to `#set()` in the `SnapController`.

File: packages/snaps-controllers/src/snaps/SnapController.ts [hunk 4]
```diff
@@ -2689,6 +2693,9 @@ export class SnapController extends BaseController<
 
       this.#updateApproval(pendingApproval.id, {
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/SnapController.ts:2694]: Please add a test or an assertion to an existing test testing that `initialConnections` is part of the `requestState`.

File: packages/snaps-controllers/src/snaps/SnapController.ts [hunk 5]
```diff
@@ -2231,6 +2231,10 @@ export class SnapController extends BaseController<
         this.#calculatePermissionsChange(snapId, processedPermissions);
 
       this.#updateApproval(pendingApproval.id, {
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/SnapController.ts:2233]: Please add a test or an assertion to an existing test testing that `initialConnections` is part of the `requestState`.

File: packages/snaps-controllers/src/snaps/SnapController.ts (hunk 2 above)
Comment (david0xd) [reply packages/snaps-controllers/src/snaps/SnapController.ts:2237]: My intention here was to omit `initialConnections` completely if it's `undefied` or `null`. I'm not sure if leaving an empty object `{}` makes sense or helps in the process? 🤔 
Also, a lot of tests would require update if I leave the empty object, it would be added everytime, everywhere related to the install/update process. WDYT?

Comment (FrederikBolding) [reply packages/snaps-controllers/src/snaps/SnapController.ts:2237]: It feels like it would be simpler to read and understand if we just always pass it. We do the same for permissions when updating as those may not always be defined either.

I know we would have to update a bunch of tests, but I think it may be worth it.

File: packages/snaps-utils/src/snaps.ts (hunk 3 above)
Comment (david0xd) [reply packages/snaps-utils/src/snaps.ts:95]: Added it to `#set()` now.

File: packages/snaps-controllers/src/snaps/SnapController.ts (hunk 4 above)
Comment (david0xd) [reply packages/snaps-controllers/src/snaps/SnapController.ts:2694]: Added.

File: packages/snaps-controllers/src/snaps/SnapController.ts (hunk 5 above)
Comment (david0xd) [reply packages/snaps-controllers/src/snaps/SnapController.ts:2233]: Added.

File: packages/snaps-controllers/src/snaps/SnapController.ts (hunk 2 above)
Comment (david0xd) [reply packages/snaps-controllers/src/snaps/SnapController.ts]: Alright, added it now as you suggested with `?? {}` and updated the tests.

File: packages/snaps-controllers/src/snaps/SnapController.ts (hunk 1 above)
Comment (david0xd) [reply packages/snaps-controllers/src/snaps/SnapController.ts]: Refactored and answered in: https://github.com/MetaMask/snaps/pull/2322#discussion_r1559458362

File: packages/snaps-controllers/src/snaps/SnapController.test.ts [hunk 6]
```diff
@@ -539,6 +539,114 @@ describe('SnapController', () => {
     await service.terminateAllSnaps();
   });
 
+  it('makes sure that requestState has been updated with initialConnections when installing a Snap', async () => {
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/SnapController.test.ts:542]:
```suggestion
  it('includes the initialConnections data in the approval requestState when installing a Snap', async () => {
```

File: packages/snaps-controllers/src/snaps/SnapController.test.ts [hunk 7]
```diff
@@ -539,6 +539,114 @@ describe('SnapController', () => {
...
+          connections: initialConnections,
+          permissions: expect.anything(),
+        },
+      },
+    );
+
+    snapController.destroy();
+  });
+
+  it('makes sure that requestState has been updated with initialConnections when updating a Snap', async () => {
```
Comment (FrederikBolding) [inline packages/snaps-controllers/src/snaps/SnapController.test.ts:589]:
```suggestion
  it('includes the initialConnections data in the approval requestState when updating a Snap', async () => {
```

**Review by FrederikBolding — APPROVED** [review]

File: packages/snaps-controllers/src/snaps/SnapController.test.ts (hunk 6 above)
Comment (david0xd) [reply packages/snaps-controllers/src/snaps/SnapController.test.ts]: Updated.

**Review by FrederikBolding — APPROVED** [review]
